### ISSUE TITLE
Dice Coefficient

### ISSUE THREAD

**Issue by lakehanne** [jocicmarko/ultrasound-nerve-segmentation#4]
Hi there,

Thanks for posting the code. I  noticed in your calculation of the DICE coefficient, you smoothed the result by `1` i.e.

```python
def dice_coef(y_true, y_pred):
    y_true_f = K.flatten(y_true)
    y_pred_f = K.flatten(y_pred)
    intersection = K.sum(y_true_f * y_pred_f)
    return (2. * intersection + smooth) / (K.sum(y_true_f) + K.sum(y_pred_f) + smooth)


def dice_coef_loss(y_true, y_pred):
    return -dice_coef(y_true, y_pred)
```
My understanding is that 	

```math
Sorensen's original formula =  2 * |X n Y| / (  |X| + |Y| )
```
My questions:  1) why did you add the smoothing value? How does it help during training?

 2.) why did you take the negative of the coefficient in Line 19?

Pardon my questions if the answer seems self-evident as I am not a user of Keras at all but it seems to me that you computed something different from the original dice coefficient. I might be wrong so I would appreciate your explanation. Thank you!

**Comment by jocicmarko** [comment]
Hey,

1) Smoothing - if there was no smoothing, imagine the most extreme case when the bottom part of the division (denominator) is 0. You would get division by 0 which is not defined and therefore would result in NaN value, making the training impossible! Adding a smooth factor sanitizes this problem.

2) Negative value because Dice coefficient is in [0,1] range, the bigger the better. However, when optimizing weights of the neural network, usually we want to find a **minimum** of the loss (objective) function. Therefore, if we take negative of Dice coefficient, we can efficiently minimize it (thus maximizing the normal, positive of Dice coefficient).

**Comment by lakehanne** [comment]
Thanks!

**Comment by mrgloom** [comment]
Why `smooth` is not just some epsilon like `eps=1e-6` ?
